diff --git a/velox/exec/Task.cpp b/velox/exec/Task.cpp
index 159f3a4a4..a216e54ff 100644
--- a/velox/exec/Task.cpp
+++ b/velox/exec/Task.cpp
@@ -1634,6 +1634,7 @@ ContinueFuture Task::terminate(TaskState terminalState) {
       return makeFinishFutureLocked("Task::terminate");
     }
     state_ = terminalState;
+    /*
     if (state_ == TaskState::kCanceled || state_ == TaskState::kAborted) {
       try {
         VELOX_FAIL(
@@ -1643,6 +1644,7 @@ ContinueFuture Task::terminate(TaskState terminalState) {
         exception_ = std::current_exception();
       }
     }
+    */
 
     LOG(INFO) << "Terminating task " << taskId() << " with state "
               << taskStateString(state_) << " after running for "
