diff --git a/velox/connectors/tpch/TpchConnector.cpp b/velox/connectors/tpch/TpchConnector.cpp
index 1bf4eac4..e365561b 100644
--- a/velox/connectors/tpch/TpchConnector.cpp
+++ b/velox/connectors/tpch/TpchConnector.cpp
@@ -129,8 +129,12 @@ void TpchDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
   currentSplit_ = std::dynamic_pointer_cast<TpchConnectorSplit>(split);
   VELOX_CHECK(currentSplit_, "Wrong type of split for TpchDataSource.");
 
+  size_t tpchTableRowCount = tpchTable_ == tpch::Table::TBL_LINEITEM
+      ? getRowCount(tpch::Table::TBL_ORDERS, scaleFactor_)
+      : tpchTableRowCount_;
+
   size_t partSize =
-      std::ceil((double)tpchTableRowCount_ / (double)currentSplit_->totalParts);
+      std::ceil((double)tpchTableRowCount / (double)currentSplit_->totalParts);
 
   splitOffset_ = partSize * currentSplit_->partNumber;
   splitEnd_ = splitOffset_ + partSize;
diff --git a/velox/tpch/gen/TpchGen.cpp b/velox/tpch/gen/TpchGen.cpp
index 4ccd3ee2..3a972bff 100644
--- a/velox/tpch/gen/TpchGen.cpp
+++ b/velox/tpch/gen/TpchGen.cpp
@@ -456,7 +456,7 @@ RowVectorPtr genTpchLineItem(
 
       lineNumberVector->set(lineItemCount + l, line.lcnt);
 
-      quantityVector->set(lineItemCount + l, decimalToDouble(line.quantity));
+      quantityVector->set(lineItemCount + l, 100 * decimalToDouble(line.quantity));
       extendedPriceVector->set(lineItemCount + l, decimalToDouble(line.eprice));
       discountVector->set(lineItemCount + l, decimalToDouble(line.discount));
       taxVector->set(lineItemCount + l, decimalToDouble(line.tax));
     